### PR TITLE
making

### PR DESCRIPTION
**Pull request by valeIT**

### CODE REVIEW

File: _posts/2019-11-28-making-software-is-hard.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+type: post
+published: true
+status: publish
+categories: [Programming]
+image:
+image2:
+author: Valentino Urbano
+---
+
+It is hard for non technical people to understand the amount of work that goes on behind the scene in creating any kind of non trivial application. It is even harder to make them understand that 'just changing x' might not be as trivial as they think it would be. Only as a software engineer yourself[^1] you can assume with some degree of confidence what would generally be the amount of work required for a certain feature.
```
Comment (MsSlacker) [inline _posts/2019-11-28-making-software-is-hard.md]: Change second trivial to easy so it doesn't repeat

File: _posts/2019-11-28-making-software-is-hard.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+I doubt we'll reach a point where the general population understands that most changes they think about are actually much harder to implement than they imagine anytime soon (if ever).
+
+Sure, you might argue that some times it is the opposite. Changing something may just mean changing one line of code when from an outside perspective it might seem like not such a trivial change, but the opposite case is 99% more common.
+
+## Estimates
+
+### Deadlines
+
+Estimates in software are notoriously hard. People underestimate work by an order of magnitude. Whatever the metodology to do estimates that we use we usually think about the new feature to develop in isolation, but we don't estimate the time that it might take for hiccups, interfacing with the rest of the system, having other priorities that limit our available time to work on such feature and so on. This results in an industry that in general keeps failing to keep its promise given at estimation time with projects constantly over budget and delayed. Public projects are notorious for this (healtcare.gov as a "fairly recent" example that got a lot of media coverage), but in private projects the same is true as well.
```
Comment (MsSlacker) [inline _posts/2019-11-28-making-software-is-hard.md]: metodology -> methodology

Comment (MsSlacker) [reply _posts/2019-11-28-making-software-is-hard.md]: And that sentence needs rewriting...

Comment (MsSlacker) [reply _posts/2019-11-28-making-software-is-hard.md]: estimation - > estimated

File: _posts/2019-11-28-making-software-is-hard.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+The real problem is that once you know that you will go over time there is not really a solution to speed up.
+
+- Crunching
+
+A lot of companies decide to force people to work overtime to try and complete a project in time. This will work for the first week, but as time goes on and people get less sleep consistently the start to perform worse and worse and you end up paying them overtime for less work than they would have done if they well working 8 hours, but well rested. This is not taking into consideration all the avverse effect on health from being overworked and not sleeping enough. Also you increase the risk of people burning out from overworking and taking productivity even further.
```
Comment (MsSlacker) [inline _posts/2019-11-28-making-software-is-hard.md]: the - > they

Comment (MsSlacker) [reply _posts/2019-11-28-making-software-is-hard.md]: They get paid? O.o

Comment (MsSlacker) [reply _posts/2019-11-28-making-software-is-hard.md]: taking - > tanking

File: _posts/2019-11-28-making-software-is-hard.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+- Hiring more people
+
+Throwing more people into the arena increases complexity so you will waste more time dealing with tooling, interpersonal relashionships, some work will be blocked by pending work from someone else and so on. It might work for certain projects, but generally you only get a 0.x increase in productivity and not the expected doubling, this factor also decreases as you add more people. If 2 people instead of one gives a 1.7 factor 3 people might give a 2.2 factor (numbers not accurate, just to give an idea).
```
Comment (MsSlacker) [inline _posts/2019-11-28-making-software-is-hard.md]: relashionships - > relationships

File: _posts/2019-11-28-making-software-is-hard.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+- Decreasing Scope
+
+The only real solution is a change in scope. If the deadline can't be changed and we also defined that we can increase productivity after a certain extent the only thing left is to cut on features. What is really necessary for the release? Can something be cut out? We've seen this clearly in this year's iOS 13. The OS was not ready for the new iPhone release so Apple decided to ship a version without some of the feature announced at WWDC in June and add them a while later with iOS 13.1
```
Comment (MsSlacker) [inline _posts/2019-11-28-making-software-is-hard.md]: feature - > features

File: _posts/2019-11-28-making-software-is-hard.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+Some times you only need to talk with the client/designers more to get some nebulous parts clarified
+
+- Some things will always be unknown until you begin
+
+It's in the nature of software that some things won't be clear until you actually start working on them. There is nothing you can do about that. What you can do is to take that into account while estimating that feature.
+
+## Conclusions
+
+I hope I gave you some tools to give better estimates. Unfotunately you have as developer not a lot of control on the design and product analysis part of the development process that might give you a clearer picture. You just need to be aware of it and act accordingly.
```
Comment (MsSlacker) [inline _posts/2019-11-28-making-software-is-hard.md]: Unfotunately you have as developer not a lot of control on - > Unfotunately you, as developer, have not a lot of control on

Comment (MsSlacker) [reply _posts/2019-11-28-making-software-is-hard.md]: On - > in?

**Review by MsSlacker — CHANGES_REQUESTED** [review]

File: _posts/2019-11-28-making-software-is-hard.md (hunk 2 above)
Comment (valeIT) [reply _posts/2019-11-28-making-software-is-hard.md]: > estimation - > estimated

given at estimation time

Don't think that ed is correct here

File: _posts/2019-11-28-making-software-is-hard.md (hunk 3 above)
Comment (valeIT) [reply _posts/2019-11-28-making-software-is-hard.md]: Can't find the - > they

File: _posts/2019-11-28-making-software-is-hard.md (hunk 6 above)
Comment (valeIT) [reply _posts/2019-11-28-making-software-is-hard.md]: mm I think on is correct there

**Comment by valeIT** [comment]
Left a few comments

**Review by MsSlacker — APPROVED** [review]

**Pull request merged**